apps/public/public_test.go

```diff
@@ -234,6 +237,169 @@ func TestAcquireTokenWithTenantID(t *testing.T) {
+
+	clientInfo := base64.RawStdEncoding.EncodeToString([]byte(`{"uid":"uid","utid":"utid"}`))
+	lmo, tenant := "login.microsoftonline.com", "tenant"
+	authority := fmt.Sprintf("https://%s/%s", lmo, tenant)
+	accessToken, idToken, refreshToken := "at", mock.GetIDToken(tenant, lmo), "rt"
+	for _, test := range []struct {
+		capabilities     []string
+		claims, expected string
+	}{
+		{},
+		{
+			capabilities: []string{"cp1"},
+			expected:     `{"access_token":{"xms_cc":{"values":["cp1"]}}}`,
+		},
+		{
+			claims:   `{"id_token":{"auth_time":{"essential":true}}}`,
+			expected: `{"id_token":{"auth_time":{"essential":true}}}`,
+		},
+		{
+			capabilities: []string{"cp1", "cp2"},
+			claims:       `{"access_token":{"nbf":{"essential":true, "value":"42"}}}`,
+			expected:     `{"access_token":{"nbf":{"essential":true, "value":"42"}, "xms_cc":{"values":["cp1","cp2"]}}}`,
+		},
+	} {
+		var expected map[string]any
+		if err := json.Unmarshal([]byte(test.expected), &expected); err != nil && test.expected != "" {
+			t.Fatal("test bug: the expected result must be JSON or an empty string")
+		}
+		// validate determines whether a request's query or form values contain the expected claims
+		validate := func(t *testing.T, v url.Values) {
+			if test.expected == "" {
+				if v.Has("claims") {
+					t.Fatal("claims shouldn't be set")
+				}
+				return
+			}
+			claims, ok := v["claims"]
+			if !ok {
+				t.Fatal("claims should be set")
+			}
+			if len(claims) != 1 {
+				t.Fatalf("expected exactly 1 claims value, got %d", len(claims))
+			}
+			var actual map[string]any
+			if err := json.Unmarshal([]byte(claims[0]), &actual); err != nil {
+				t.Fatal(err)
+			}
+			if diff := pretty.Compare(expected, actual); diff != "" {
+				t.Fatal(diff)
+			}
+		}
+		for _, method := range []string{"authcode", "authcodeURL", "devicecode", "interactive", "password", "passwordFederated"} {
+			t.Run(method, func(t *testing.T) {
+				mockClient := mock.Client{}
+				if method == "obo" {
+					// TODO: OBO does instance discovery twice before first token request https://github.com/AzureAD/microsoft-authentication-library-for-go/issues/351
+					mockClient.AppendResponse(mock.WithBody(mock.GetInstanceDiscoveryBody(lmo, tenant)))
+				}
+				mockClient.AppendResponse(mock.WithBody(mock.GetInstanceDiscoveryBody(lmo, tenant)))
+				mockClient.AppendResponse(mock.WithBody(mock.GetTenantDiscoveryBody(lmo, tenant)))
+				switch method {
+				case "devicecode":
+					mockClient.AppendResponse(mock.WithBody([]byte(`{"device_code":".","expires_in":600}`)))
+				case "password":
+					mockClient.AppendResponse(mock.WithBody([]byte(`{"account_type":"Managed","cloud_audience_urn":".","cloud_instance_name":".","domain_name":"."}`)))
+				case "passwordFederated":
+					mockClient.AppendResponse(mock.WithBody([]byte(`{"account_type":"Federated","cloud_audience_urn":".","cloud_instance_name":".","domain_name":".","federation_protocol":".","federation_metadata_url":"."}`)))
```

Comment:
Outside of the scope of this PR, but a FYI: [MSALs are allowing confidential clients to use password flow (internal work item)](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/2092736).